File: Year-2019/Mar/2019-03-13-152/lb/152.java
class Solution {
    /**
     * 多了一层反转的情况，需要记录最小值
     * @param nums
     * @return
     */
    public int maxProduct(int[] nums) {
        if (nums.length == 0)
            return 0;
        int max=1;
        int min=1;
        int ret=nums[0];
        int temp;
        for(int i =0;i<nums.length;i++){
            //小于零则会产生大变小，小变大的情况
            if (nums[i]<0){
                temp = max;
                max = min;
                min = temp;
            }
            max = Math.max(max*nums[i],nums[i]);//记录中间产生的最大值，比较前n个数的乘积和当前数的大小，若小于说明该换区间了，发生在负数之后
            min = Math.min(min*nums[i],nums[i]);//伺机而动，等着翻身做主人
            ret = Math.max(max,ret);//整个过程中的最大值
        }
        return ret;
    }
}

File: Year-2019/Mar/2019-03-13-152/watercow/152.c
class Solution {
public:
	int maxProduct(vector<int>& nums) {
		if (nums.size() == 1) return nums[0];

		int len = nums.size();
		vector<int> dp_max(len, 0), dp_min(len, 0);
		int result = INT_MIN;

		for (int i = 0; i < len; i++) {
			if (i == 0) {
				dp_max[i] = nums[i];
				dp_min[i] = nums[i];
			}
			else {
				dp_max[i] = max(dp_max[i - 1] * nums[i], max(dp_min[i - 1] * nums[i], nums[i]));
				dp_min[i] = min(dp_max[i - 1] * nums[i], min(dp_min[i - 1] * nums[i], nums[i]));
			}
			result = max(dp_max[i], result);
		}
		return result;
	}
};